include/deal.II/lac/householder.h

```diff
@@ -190,7 +189,11 @@ Householder<number>::initialize(const FullMatrix<number2> &M)
       if (std::fabs(sigma) < 1.e-15)
         return;
 
-      number2 s = (storage(j, j) < 0) ? std::sqrt(sigma) : -std::sqrt(sigma);
+      number2 s;
+      if constexpr (numbers::NumberTraits<number2>::is_complex)
+        s = storage(j, j).real() < 0 ? std::sqrt(sigma) : -std::sqrt(sigma);
```

Comment:
Is there some citation for this algorithm? I'm curious about why we need this check.

Reply:
This is a good question. It is usually done to avoid cancellation errors in the computation that does `sigma - s * storage(j, j)` below by ensuring the same sign on both terms. You can find this e.g. on Wikipedia: https://en.wikipedia.org/wiki/Householder_transformation (look out for the text immediately after "catastrophic cancellation").